feat(cd): sign released container images and Helm chart

hainenber opened the pull request:
# Description

Sign released container images with Cosign and Helm chart with `helm package --sign`

## Related Issue

If this pull request is related to any issue, please mention it here. Additionally, make sure that the issue is assigned to you before submitting this pull request.

Closes #141 

## Checklist

- [x] I have read the [contributing documantation](https://retina.sh/docs/contributing).
- [x] I signed and signed-off the commits (`git commit -S -s ...`)
- [x] I have correctly attributed the author(s) of the code.
- [ ] I have tested the changes locally.
- [ ] I have followed the project's style guidelines.
- [ ] I have updated the documentation, if necessary.
- [ ] I have added tests, if applicable.

## Screenshots (if applicable)

N/A

## Additional Notes

In order for testing this change, please create a GPG private key and a Cosign private key and deposit as GH secrets for this repo

---

Please refer to the [CONTRIBUTING.md](../CONTRIBUTING.md) file for more information on how to contribute to this project.

rbtr commented:
@hainenber thanks for picking this up, I'll give this a thorough review ASAP.

One initial question - I noticed in the [GH sample workflow they don't have any explicit keys](https://github.com/actions/starter-workflows/blob/main/ci/docker-publish.yml#L90-L98). I think this is because they are using [OIDC integration to do "keyless" signing](https://github.blog/2021-12-06-safeguard-container-signing-capability-actions/#:~:text=This%20means%20you,third%2Dparty%20service.)?

This seems preferable since we wouldn't need to manage GPG keys as secrets. Do you think that's something we can do here?

hainenber commented:
Sure, I think it's feasible if GH decided to demonstrate their sample workflow as such. Will tinker around a bit

rbtr commented on .github/workflows/goreleaser.yaml:
```diff
@@ -48,3 +52,12 @@ jobs:
           args: release --clean
         env:
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+      - name: Sign container image
+        run: |
+          export TAG=(make version)
+          cosign sign --key cosign.key \
+            ghcr.io/${{ github.repository }}/kubectl-retina:$TAG \
+            ghcr.io/${{ github.repository }}/kubectl-retina:$TAG-linux-amd64 \
+            ghcr.io/${{ github.repository }}/kubectl-retina:$TAG-linux-arm64
+        env:
+          COSIGN_PASSWORD: "" 
```
I don't think we need cosign in the GoReleaser job at all - it is only used to produce binaries for kubectl-retina. Attaching binaries to the releases supersedes the kubectl-retina image and we don't build or publish it anymore.

hainenber replied:
Okie, will remove this in the next commit

hainenber replied:
Removed :D

hainenber commented:
All relevant container images and Helm chart are now signed with `cosign` in keyless manner. There's 1 caveat with signing Helm chart though; it'll generate a `cosign.bundle` - a base64-encoded string containing both certificate and signature.

I decide to make it into a build artifact so release engineers can bundle (no pun, I swear) into a release if needed.

rbtr commented on .github/workflows/release-charts.yaml:
```diff
@@ -22,15 +25,28 @@ jobs:
+          echo "TAG=$TAG" >> $GITHUB_OUTPUT
           echo $TAG
           helm package ./deploy/manifests/controller/helm/retina --version $TAG
+          cosign sign-blob --yes retina-$TAG.tgz --bundle retina-$TAG.cosign.bundle
```
Maybe this way is better, but I'm trying to understand: why sign the bundle, vs signing the OCI image after it is pushed similar to the container steps in release-images?
It seems like we could 
```bash 
DIGEST=$(docker inspect --format='{{index .RepoDigests 0}}' ghcr.io/${{ github.repository }}/charts/retina:$TAG)
cosign sign --yes ${DIGEST}
```
in the same fashion?

rbtr commented on .github/workflows/release-images.yaml:
```diff
@@ -43,6 +49,18 @@ jobs:
+      - name: Sign container image
+        run: |
+          for image in \
+            ghcr.io/${{ github.repository }}/retina-agent:$TAG \
+            ghcr.io/${{ github.repository }}/retina-agent:$TAG-${{ matrix.platform }}-${{ matrix.arch }} \
+            ghcr.io/${{ github.repository }}/retina-init:$TAG \ 
+            ghcr.io/${{ github.repository }}/retina-init:$TAG-${{ matrix.platform }}-${{ matrix.arch }} \
```
```suggestion
            ghcr.io/${{ github.repository }}/retina-agent:$TAG-${{ matrix.platform }}-${{ matrix.arch }} \
            ghcr.io/${{ github.repository }}/retina-init:$TAG-${{ matrix.platform }}-${{ matrix.arch }} \
```
The multi-platform images are created during the `manifests` job, they are not yet available here where we are just building all of the platform-specific images separately.

rbtr commented on .github/workflows/release-images.yaml:
```diff
@@ -111,6 +147,16 @@ jobs:
+      - name: Sign container image
+        run: |
+          for image in \
+            ghcr.io/${{ github.repository }}/retina-operator:$TAG \
+            ghcr.io/${{ github.repository }}/retina-operator:$TAG-${{ matrix.platform }}-${{ matrix.arch }} \
```
```suggestion
            ghcr.io/${{ github.repository }}/retina-operator:$TAG-${{ matrix.platform }}-${{ matrix.arch }} \
```
same comment re: manifests

rbtr commented on .github/workflows/release-images.yaml:
```diff
@@ -111,6 +147,16 @@ jobs:
+            cosign sign --yes ${DIGEST}
+          done
 
   manifests:
```
Will we need to sign the multiplat manifests during this job also?

hainenber replied:
Well yes, that's where the images without `${os}-${arch}` got generated and pushed. Thanks for the callout!

hainenber replied:
This is added in HEAD

rbtr reviewed (changes requested):
This is great, thanks! Left some feedback based on my testing

rbtr commented:
I did some experimenting with this and I couldn't quite get it to work: https://github.com/rbtr/retina/actions/runs/8474428281
Here are the changes I made: https://github.com/rbtr/retina/commit/04c7fe61670c2360ddef030054ad5e099ebbde1b

Ran in to a few problems:
- the signing step uses the image digest. We don't have that image loaded in `docker`, it's in the `buildx` context, so we have to pull it or get the digest earlier and save it. If we try to pull it, we run in to platform issues - it's tricky to pull an image for a platform that's different from the host, especially if it's Windows.
- I can't figure out the command to verify a signature...it should be something like `cosign verify ghcr.io/rbtr/retina/retina-agent@sha256:1d38a1d76f032a95ea502ec1d52dc2fc1475aa18c6a6c9ed41841c9a7fe2f0d8 --certificate-identity "https://github.com/rbtr/retina" --certificate-oidc-issuer "https://token.actions.githubusercontent.com"` but I just get `Error: no matching signatures`. I think the certificate-identity is wrong? But when I use certificate-identity-regexp "*", still nothing.

Overall I think this is very close with the changes suggested earlier, if we can figure out a solution to the digest problem. For acceptance criteria I would love to see a published signed image (and be able to verify it with cosign) to demonstrate that this functions 🙂

hainenber commented:
> I did some experimenting with this and I couldn't quite get it to work: https://github.com/rbtr/retina/actions/runs/8474428281 Here are the changes I made: [rbtr@04c7fe6](https://github.com/rbtr/retina/commit/04c7fe61670c2360ddef030054ad5e099ebbde1b)
> 
> Ran in to a few problems:
> 
> * the signing step uses the image digest. We don't have that image loaded in `docker`, it's in the `buildx` context, so we have to pull it or get the digest earlier and save it. If we try to pull it, we run in to platform issues - it's tricky to pull an image for a platform that's different from the host, especially if it's Windows.
> * I can't figure out the command to verify a signature...it should be something like `cosign verify ghcr.io/rbtr/retina/retina-agent@sha256:1d38a1d76f032a95ea502ec1d52dc2fc1475aa18c6a6c9ed41841c9a7fe2f0d8 --certificate-identity "https://github.com/rbtr/retina" --certificate-oidc-issuer "https://token.actions.githubusercontent.com"` but I just get `Error: no matching signatures`. I think the certificate-identity is wrong? But when I use certificate-identity-regexp "*", still nothing.
> 
> Overall I think this is very close with the changes suggested earlier, if we can figure out a solution to the digest problem. For acceptance criteria I would love to see a published signed image (and be able to verify it with cosign) to demonstrate that this functions 🙂

Thanks for your effort on testing out the changes on your forked repo! My apologies for having this under the radar as I tend to disable GH actions in my forked repos to cut CI costs

Re: 1st point, I'm trying to store image digest with this [`--metadata-file`](https://github.com/docker/buildx/blob/master/docs/reference/buildx_build.md#metadata-file). This should be able to cover this concern once I'm done with another round of tinkering.

Onto the 2nd point, from [`cosign` docs](https://docs.sigstore.dev/verifying/verify/#keyless-verification-using-openid-connect), it seems the OIDC issuer for GH-stored images is "https://github.com/login/oauth" and the cert's identity is the signer's address. Another round of tinkering is needed to validate this ofc.

Onto the lab, hopefully a short round this time.

To save up your effort to other Retina's more pressuring issues, I'll convert this one to Draft and will seek your reviews once it deems as worthy.

rbtr commented:
> Thanks for your effort on testing out the changes on your forked repo! My apologies for having this under the radar as I tend to disable GH actions in my forked repos to cut CI costs
> 
> Re: 1st point, I'm trying to store image digest with this [`--metadata-file`](https://github.com/docker/buildx/blob/master/docs/reference/buildx_build.md#metadata-file). This should be able to cover this concern once I'm done with another round of tinkering.
> 
> Onto the 2nd point, from [`cosign` docs](https://docs.sigstore.dev/verifying/verify/#keyless-verification-using-openid-connect), it seems the OIDC issuer for GH-stored images is "https://github.com/login/oauth" and the cert's identity is the signer's address. Another round of tinkering is needed to validate this ofc.
> 
> Onto the lab, hopefully a short round this time.
> 
> To save up your effort to other Retina's more pressuring issues, I'll convert this one to Draft and will seek your reviews once it deems as worthy.

Happy to collab on this, no problem at all. This is an important feature to me :)
I can keep pulling it into my fork or you can PR there to get Actions runs? Let me know how I can help here.

rbtr commented:
thanks @hainenber! I will test this out 🙂

hainenber commented:
Please do! Bon apetit :D

nddq reviewed (commented):
left some nit comments, thanks!

rbtr reviewed (dismissed):
Functionally LGTM! This is great, thanks @hainenber 🍻
Aside from anyone else's comments, my last ask would be to updated the README and docs.
I'm thinking we add a block such as:
```markdown
### Verify signed images
Retina images published to GHCR are cryptographically signed. You can verify their provenance with [`sigstore/cosign`](https://github.com/sigstore/cosign):

```shell
REPO=microsoft/retina # or your repo
IMAGE=retina-operator # or other image to verify
TAG=v0.0.6 # or other tag to verify OR replace with the image SHA256
cosign verify ghcr.io/$REPO/$IMAGE:$TAG --certificate-oidc-issuer https://token.actions.githubusercontent.com --certificate-identity-regexp="https://github.com/$REPO" -o text
\```
```

to the README, and duplicate it to a site page such as `docs/installation/verify-signed-images.md` so that it's on retina.sh also 🙂

hainenber commented on .github/workflows/release-images.yaml:
```diff
@@ -78,6 +96,16 @@ jobs:
+      - name: Sign container image
+        run: |
+          for year in 2019 2022; do
+            for image in retina-agent ; do
```
I think it'd make extending the signing process easier since there's already a loop

hainenber commented on .github/workflows/release-images.yaml:
```diff
@@ -111,6 +142,15 @@ jobs:
+          
+      - name: Sign container image
+        run: |
+          for image in retina-operator ; do
```
same logic as above too

hainenber commented:
@rbtr I've addressed Quang's comments and added relevant documentation as suggested by you. I think yours is already succinct so pardon me for a whole copy-paste :D

rbtr commented on .github/workflows/release-images.yaml:
```diff
@@ -138,3 +181,16 @@ jobs:
+            IMAGE_PATH="ghcr.io/${{ github.repository }}/$image:$TAG"
+            DIGEST=$(docker buildx imagetools inspect $IMAGE_PATH --format "{{json .Manifest}}" | jq -r .digest)
+            cosign sign --yes ${IMAGE_PATH}@${DIGEST} 
+          done
```
```suggestion
          done

```

rbtr commented on docs/installation/verify-signed-images.md:
````diff
@@ -0,0 +1,10 @@
+IMAGE=retina-operator # or other image to verify
+TAG=v0.0.6 # or other tag to verify OR replace with the image SHA256
+cosign verify ghcr.io/$REPO/$IMAGE:$TAG --certificate-oidc-issuer https://token.actions.githubusercontent.com --certificate-identity-regexp="https://github.com/$REPO" -o text
+```
````
````suggestion
```

````

rbtr reviewed (approved):
@hainenber I committed newlines at the end of your markdown files to make the linter happy, everything else LGTM

rbtr commented on .github/workflows/release-charts.yaml:
```diff
@@ -22,15 +25,25 @@ jobs:
           helm package ./deploy/manifests/controller/helm/retina --version $TAG
-          helm push retina-$TAG.tgz oci://ghcr.io/${{ github.repository }}/charts
+          helm push retina-$TAG.tgz oci://ghcr.io/${{ github.repository }}/charts >> helm_push_result.txt 2>&1
+          cat helm_push_result.txt
```
maybe clear enough in the `helm push` docs? 
https://helm.sh/docs/topics/registries/#the-push-subcommand
